src/Stratis.Bitcoin/FinalizedBlockInfoRepository.cs

```diff
@@ -0,0 +1,208 @@
+    /// <remarks>
+    /// Finalized block height is the height of the last block that can't be reorged.
+    /// Blocks with height greater than finalized height can be reorged.
+    /// <para>Finalized block height value is always <c>0</c> for blockchains without max reorg property.</para>
+    /// </remarks>
+    public interface IFinalizedBlockInfoRepository : IDisposable
+    {
+        /// <summary>Gets the finalized block hash and height.</summary>
+        /// <returns>Hash and height of a block that can't be reorged away from.</returns>
+        HashHeightPair GetFinalizedBlockInfo();
+
+        /// <summary>Loads the finalised block hash and height from the database.</summary>
+        Task LoadFinalizedBlockInfoAsync(Network network);
+
+        /// <summary>Saves the finalized block hash and height to the database if height is greater than the previous value.</summary>
+        /// <param name="hash">Block hash.</param>
+        /// <param name="height">Block height.</param>
+        /// <returns><c>true</c> if new value was set, <c>false</c> if <paramref name="height"/> is lower or equal than current value.</returns>
+        bool SaveFinalizedBlockHashAndHeight(uint256 hash, int height);
+    }
+
+    public class FinalizedBlockInfoRepository : IFinalizedBlockInfoRepository
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Access to DBreeze database.</summary>
+        private readonly DBreezeEngine dbreeze;
+
+        /// <summary>Database key under which the block height of the last finalized block height is stored.</summary>
+        private static readonly byte[] finalizedBlockKey = new byte[0];
+
+        /// <summary>Height and hash of a block that can't be reorged away from.</summary>
+        private HashHeightPair finalizedBlockInfo;
+
+        /// <summary>Queue of finalized infos to save.</summary>
+        /// <remarks>All access should be protected by <see cref="queueLock"/></remarks>
+        private readonly Queue<HashHeightPair> finalizedBlockInfosToSave;
+
+        /// <summary>Protects access to <see cref="finalizedBlockInfosToSave"/>.</summary>
+        private readonly object queueLock;
+
+        /// <summary>Task that continously persists finalized block info to the database.</summary>
+        private readonly Task finalizedBlockInfoPersistingTask;
+
+        private readonly CancellationTokenSource cancellation;
+
+        private readonly AsyncManualResetEvent queueUpdatedEvent;
+
+        public FinalizedBlockInfoRepository(string folder, ILoggerFactory loggerFactory)
+        {
+            Guard.NotEmpty(folder, nameof(folder));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            Directory.CreateDirectory(folder);
+            this.dbreeze = new DBreezeEngine(folder);
```

Comment:
One think I don't understand is why create a new store we could like before just reuse chain store (this will create a new folder and the whole dbreeze structure just to store a hash and a height.

Reply:
Yeah, that's for logical separation. 
We can have chain store on any chain, but finalized block is only stratis thing.

Reply:
yes but in that case it will just be empty no big deal, better then creating a whole new store just for a hash.